src/main/java/com/google/api/gax/bundling/FlowController.java

```diff
@@ -0,0 +1,187 @@
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.bundling;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+import java.util.concurrent.Semaphore;
+import javax.annotation.Nullable;
+
+/** Provides flow control capability. */
+public class FlowController {
+  /** Base exception that signals a flow control state. */
+  public abstract class FlowControlException extends Exception {
+    private FlowControlException() {}
+  }
+
+  /**
+   * Returned as a future exception when client-side flow control is enforced based on the maximum
+   * number of outstanding in-memory elements.
+   */
+  public final class ElementCountMaxOutstandingReachedException extends FlowControlException {
+    private final int currentMaxElementCount;
+
+    public ElementCountMaxOutstandingReachedException(int currentMaxElementCount) {
+      this.currentMaxElementCount = currentMaxElementCount;
+    }
+
+    public int getCurrentMaxBundleElementCount() {
+      return currentMaxElementCount;
+    }
+
+    @Override
+    public String toString() {
+      return String.format(
+          "The maximum number of bundle elements: %d have been reached.", currentMaxElementCount);
+    }
+  }
+
+  /**
+   * Returned as a future exception when client-side flow control is enforced based on the maximum
+   * number of unacknowledged in-memory bytes.
+   */
+  public final class RequestByteMaxOutstandingReachedException extends FlowControlException {
+    private final int currentMaxBytes;
+
+    public RequestByteMaxOutstandingReachedException(int currentMaxBytes) {
+      this.currentMaxBytes = currentMaxBytes;
+    }
+
+    public int getCurrentMaxBundleBytes() {
+      return currentMaxBytes;
+    }
+
+    @Override
+    public String toString() {
+      return String.format(
+          "The maximum number of bundle bytes: %d have been reached.", currentMaxBytes);
+    }
+  }
+
+  @AutoValue
+  public abstract static class Settings {
+    public static Settings DEFAULT =
+        newBuilder()
+            .setRequestByteMaxOutstanding(Optional.<Integer>absent())
+            .setElementCountMaxOutstanding(Optional.<Integer>absent())
+            .build();
+
+    /** Maximum number of outstanding elements to keep in memory before enforcing flow control. */
+    public abstract Optional<Integer> getElementCountMaxOutstanding();
+
+    /** Maximum number of outstanding bytes to keep in memory before enforcing flow control. */
+    public abstract Optional<Integer> getRequestByteMaxOutstanding();
```

Comment:
I think `getMaxOutstandingElementCount` and `getMaxOutstandingRequestBytes` would be more natural. (Note also the pluralization of `byte` -> `bytes`). (Other locations using this terminology would need to match)